### PR TITLE
ARROW-17868: [C++][Python] Restore the ARROW_PYTHON CMake option

### PR DESCRIPTION
**Pull request by kou**
Restore it but it's marked as a deprecated option. Because the Python component in Apache Arrow C++ was moved to PyArrow by ARROW-16340. It' removed in a feature release.

Users should use CMake presets instead of ARROW_PYTHON but CMake presets requires CMake 3.19 or later.

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-release -g nightly-packaging

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-release -g nightly-packaging

File: ci/docker/linux-apt-r.dockerfile [hunk 1]
```diff
@@ -103,13 +103,18 @@ ENV \
...
     ARROW_BUILD_TESTS=OFF \
     ARROW_BUILD_UTILITIES=OFF \
+    ARROW_COMPUTE=ON \
+    ARROW_CSV=ON \
+    ARROW_DATASET=ON \
+    ARROW_FILESYSTEM=ON \
     ARROW_FLIGHT=OFF \
     ARROW_GANDIVA=OFF \
+    ARROW_HDFS=ON \
+    ARROW_JSON=ON \
```
Comment (pitrou) [inline ci/docker/linux-apt-r.dockerfile:113]: Is all this required for R? AFAIR, Python is only used to test PyArrow-R interoperability here.

File: docs/source/developers/cpp/building.rst [hunk 2]
```diff
@@ -338,11 +338,11 @@ boolean flags to ``cmake``.
...
 * ``-DARROW_PLASMA=ON``: Plasma Shared Memory Object Store
 * ``-DARROW_PLASMA_JAVA_CLIENT=ON``: Build Java client for Plasma
-* ``-DARROW_PYTHON=ON``: Arrow Python C++ integration library (required for
-  building pyarrow). This library must be built against the same Python version
-  for which you are building pyarrow. NumPy must also be installed. Enabling
-  this option also enables ``ARROW_COMPUTE``, ``ARROW_CSV``, ``ARROW_DATASET``,
-  ``ARROW_FILESYSTEM``, ``ARROW_HDFS``, and ``ARROW_JSON``.
+* ``-DARROW_PYTHON=ON``: This option is deprecated since 10.0.0. This
+  will be removed in a future release. Use CMake presets instead. Or
+  you can use ``ARROW_COMPUTE``, ``ARROW_CSV``, ``ARROW_DATASET``,
```
Comment (pitrou) [inline docs/source/developers/cpp/building.rst:343]:
```suggestion
  you can enable ``ARROW_COMPUTE``, ``ARROW_CSV``, ``ARROW_DATASET``,
```

File: ci/scripts/python_wheel_macos_build.sh [hunk 3]
```diff
@@ -96,25 +96,27 @@ cmake \
     -DARROW_BUILD_SHARED=ON \
     -DARROW_BUILD_STATIC=OFF \
     -DARROW_BUILD_TESTS=OFF \
+    -DARROW_COMPUTE=ON \
+    -DARROW_CSV=ON \
     -DARROW_DATASET=${ARROW_DATASET} \
```
Comment (jorisvandenbossche) [inline ci/scripts/python_wheel_macos_build.sh:101]: Should we just hardcode this to ON? (the ARROW_PYTHON=ON would have ensure that) I don't think we want to create wheels without dataset enabled?

File: ci/scripts/python_wheel_macos_build.sh [hunk 4]
```diff
@@ -129,9 +131,9 @@ cmake \
     -DCMAKE_INSTALL_PREFIX=${build_dir}/install \
     -DCMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES} \
     -DCMAKE_UNITY_BUILD=${CMAKE_UNITY_BUILD} \
-    -DOPENSSL_USE_STATIC_LIBS=ON \
```
Comment (jorisvandenbossche) [inline ci/scripts/python_wheel_macos_build.sh:132]: This seems unrelated to the other changes. Can you give a short reasoning for it?

File: ci/scripts/python_wheel_manylinux_build.sh [hunk 5]
```diff
@@ -89,31 +89,32 @@ pushd /tmp/arrow-build
...
 # ships the fix.
 cmake \
-    -DARROW_BROTLI_USE_SHARED=OFF \
     -DARROW_BUILD_SHARED=ON \
     -DARROW_BUILD_STATIC=OFF \
     -DARROW_BUILD_TESTS=OFF \
+    -DARROW_COMPUTE=ON \
+    -DARROW_CSV=ON \
     -DARROW_DATASET=${ARROW_DATASET} \
+    -DARROW_DATASET=ON \
```
Comment (jorisvandenbossche) [inline ci/scripts/python_wheel_manylinux_build.sh:98]:
```suggestion
    -DARROW_DATASET=ON \
```

File: ci/docker/linux-apt-r.dockerfile (hunk 1 above)
Comment (kou) [reply ci/docker/linux-apt-r.dockerfile:113]: @nealrichardson Do you know what components are needed here?

File: cpp/CMakePresets.json [hunk 6]
```diff
@@ -117,17 +117,46 @@
...
+      "name": "features-python-minimal",
+      "inherits": [
+	  "features-minimal"
+      ],
+      "hidden": true,
+      "cacheVariables": {
+        "ARROW_COMPUTE": "ON",
+        "ARROW_CSV": "ON",
+        "ARROW_FILESYSTEM": "ON",
+        "ARROW_JSON": "ON"
```
Comment (jorisvandenbossche) [inline cpp/CMakePresets.json:130]: I haven't yet used the presets myself, but shall we include `ARROW_DATASET` here as well? (that was included with the previous `features-python` preset. 
Maybe you can also keep the exact name to keep it working for people that were using those presets?

cc @wjones127

File: docs/source/developers/cpp/building.rst (hunk 2 above)
Comment (kou) [reply docs/source/developers/cpp/building.rst]: Thanks!

File: docs/source/developers/python.rst [hunk 7]
```diff
@@ -313,24 +313,26 @@ created above (stored in ``$ARROW_HOME``):
...
-           -DARROW_PARQUET=ON \
+           -DARROW_WITH_ZLIB=ON \
+           -DARROW_WITH_ZSTD=ON \
            -DPARQUET_REQUIRE_ENCRYPTION=ON \
-           -DARROW_PYTHON=ON \
-           -DARROW_BUILD_TESTS=ON \
            ..
-   $ make -j4
-   $ make install
+   $ cmake --build . --target install --config Debug
```
Comment (jorisvandenbossche) [inline docs/source/developers/python.rst:335]: For my understanding: is the `--config Debug` needed if you already use `CMAKE_BUILD_TYPE=Debug` above? 

And will this line also build in parallel?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks! Added a few comments

File: cpp/CMakePresets.json (hunk 6 above)
Comment (wjones127) [reply cpp/CMakePresets.json:130]: I think I'm fine not including datasets if it's not required for the Python build. In most workflows with CMakePresets.json, I think we expect developers to create their own presets in CMakeUserPresets.json, which will inherit from the provided once. (See [an example here](https://www.datawill.io/posts/apache-arrow-development-cpp-conda-macos/#cmakeuserpresets.json))

Once this is merged, I can send a notice to the mailing list with instructions on how to transition to the new presets. (I need to update my blog post anyways.)

File: docs/source/developers/python.rst (hunk 7 above)
Comment (wjones127) [reply docs/source/developers/python.rst:335]: Yeah the `--config Debug` is redundant.

In my experience, it seems to automatically build in parallel, but you can also [add the option explicitly](https://cmake.org/cmake/help/latest/release/3.12.html#command-line).

File: ci/scripts/python_wheel_macos_build.sh (hunk 3 above)
Comment (kou) [reply ci/scripts/python_wheel_macos_build.sh:101]: > I don't think we want to create wheels without dataset enabled

I think so but `${ARROW_DATASET}` is also used for `export PYARROW_WITH_DATASET=${ARROW_DATASET}` below. So I think that we use `${ARROW_DATASET}` here for consistency. (`ARROW_DATASET` is initialized by `: ${ARROW_DATASET:=ON}`.)

File: ci/scripts/python_wheel_macos_build.sh (hunk 4 above)
Comment (kou) [reply ci/scripts/python_wheel_macos_build.sh:132]: Yes. This is unrelated to `ARROW_PYTHON`. Sorry for including this to this pull request.

`OPENSSL_USE_STATIC_LIBS` is redundant here because there is `-DARROW_DEPENDENCY_USE_SHARED=OFF` in this command line. `OPENSSL_USE_STATIC_LIBS` is set automatically in https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindOpenSSLAlt.cmake#L48-L52 .

I just found this by sorting CMake options. So I mix this change into this pull request. Sorry.

File: ci/scripts/python_wheel_manylinux_build.sh (hunk 5 above)
Comment (kou) [reply ci/scripts/python_wheel_manylinux_build.sh]: Thanks but I choose `-DARROW_DATASET=${ARROW_DATASET}` for consistency.

File: cpp/CMakePresets.json (hunk 6 above)
Comment (kou) [reply cpp/CMakePresets.json:130]: > Maybe you can also keep the exact name to keep it working for people that were using those presets?

It makes sense. I added no `-minimal`/`-maximal` versions.

File: docs/source/developers/python.rst (hunk 7 above)
Comment (kou) [reply docs/source/developers/python.rst]: Ah, sorry. I added this accidentally. I revert this.

I found `-j4` isn't suitable here. We can use `-j$(nproc)` on Linux and `-j$(sysctl -n hw.ncpu)` on macOS. But I thought that it's better that we use Ninja here. And I started rewriting this for Ninja but I stopped it because this pull request isn't for the change. But this change remained.

File: ci/docker/linux-apt-r.dockerfile (hunk 1 above)
Comment (kou) [reply ci/docker/linux-apt-r.dockerfile:113]: It seems that we can disable `ARROW_HDFS`. Other components seems necessary. I'll try it.

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-release -g nightly-packaging

File: ci/scripts/python_wheel_macos_build.sh (hunk 4 above)
Comment (jorisvandenbossche) [reply ci/scripts/python_wheel_macos_build.sh:132]: No need to be sorry. I was just wondering about the reasoning

**Comment by kou** [comment]
I think that this is ready to merge.
Build failures aren't related to this change.
